src/error-table.rkt

```diff
@@ -1,20 +1,365 @@
-      (match-define (cons worst-subexpr pt-max-error)
-        (argmax cdr (map cons subexprs pt-errors)))
-      (if (> pt-max-error 2) worst-subexpr #f)))
+    (foldr append
+           '()
+           (for/list ([pt-errors (in-list pt-errorss)]
+                      [(pt _) (in-pcontext pcontext)])
+             (define sub-error (map cons subexprs pt-errors))
+             (define filtered-sub-error (filter (lambda (p) (> (cdr p) 16))
+                                                sub-error))
+             (define mapped-sub-error (map (lambda (p) (cons (car p) pt))
+                                           filtered-sub-error))
+             (if (empty? mapped-sub-error)
+                 (list (cons #f pt))
+                 mapped-sub-error))))
+
+  (for/hash ([group (in-list (group-by car pt-worst-subexprs))])
+    (let ([key (caar group)])
+      (values key (map cdr group)))))
+
+(define is-inexact? list?)
+(define (same-sign? a b)
+  (or (and (positive? a) (positive? b))
+      (and (negative? a) (negative? b))))
+ 
+(define (predicted-errors expr ctx pctx)
+  (define subexprs
+    (all-subexpressions-rev expr (context-repr ctx)))
+  (define subexprs-list (map car subexprs))
+ 
+  (define ctx-list
+    (for/list ([subexpr (in-list subexprs)])
+      (struct-copy context ctx [repr (cdr subexpr)])))
+ 
+  (define subexprs-fn (eval-progs-real subexprs-list ctx-list))
+ 
+  (define error-count-hash
+    (make-hash (map (lambda (x) (cons x '())) (cons #f subexprs-list))))
+ 
+  (define (mark-erroneous! expr pt)
+    (hash-update! error-count-hash expr (lambda (x) (set-add x pt))))
+  
+  (for ([(pt _) (in-pcontext pctx)])
+    (define exacts (apply subexprs-fn pt))
+    (define exacts-hash
+      (make-immutable-hash (map cons subexprs-list exacts)))
+ 
+    (define uflow-hash (make-hash (map (lambda (x) (cons x #f)) subexprs-list)))
+    (define (underflow? subexpr) (hash-ref uflow-hash subexpr))
+    (define oflow-hash (make-hash (map (lambda (x) (cons x #f)) subexprs-list)))
+    (define (overflow? subexpr) (hash-ref oflow-hash subexpr))
+ 
+    (unless
+        (ormap identity
+               (for/list ([subexpr subexprs-list])
+                 (define subexpr-val (hash-ref exacts-hash subexpr))
+
+                 (cond
+                   ; NOTE need better way to see if a calculation underflowed
+                   [(fl= subexpr-val 0.0)
+                    (hash-set! uflow-hash subexpr #t)]
+                   [(infinite? subexpr-val)
+                    (hash-set! oflow-hash subexpr #t)])
+                 
+                 (match subexpr
+                   [(list '+.f64 larg rarg)
+                    #:when (or (is-inexact? larg) (is-inexact? rarg))
+                    (define larg-val (hash-ref exacts-hash larg))
+                    (define rarg-val (hash-ref exacts-hash rarg))
+                    (define x+y (+ larg-val rarg-val))
+                    (define cond-x (abs (/ larg-val x+y)))
```

Comment:
What if denominator is 0?

Reply:
I think this is covered by the hallucination case: R(x + y) underflowed, there is no real error, but the condition number is high.

Reply:
I still don't have a way to check if its exactly 0. Might want to code that in soon too.